# Question 1: division

Question:  
Write a program which will find all such numbers which are divisible by 7 but are not a multiple of 5, between 2000 and 3200 (both included).The numbers obtained should be printed in a comma-separated sequence on a single line.

Hints:  
Consider use range(#begin, #end) method.

In [2]:
res = []
for i in range(2000,3201):
    if i % 7 == 0 and i % 5 != 0:
        print(i,end=',') #以逗号结尾，不换行
print('\b')#退格

2002,2009,2016,2023,2037,2044,2051,2058,2072,2079,2086,2093,2107,2114,2121,2128,2142,2149,2156,2163,2177,2184,2191,2198,2212,2219,2226,2233,2247,2254,2261,2268,2282,2289,2296,2303,2317,2324,2331,2338,2352,2359,2366,2373,2387,2394,2401,2408,2422,2429,2436,2443,2457,2464,2471,2478,2492,2499,2506,2513,2527,2534,2541,2548,2562,2569,2576,2583,2597,2604,2611,2618,2632,2639,2646,2653,2667,2674,2681,2688,2702,2709,2716,2723,2737,2744,2751,2758,2772,2779,2786,2793,2807,2814,2821,2828,2842,2849,2856,2863,2877,2884,2891,2898,2912,2919,2926,2933,2947,2954,2961,2968,2982,2989,2996,3003,3017,3024,3031,3038,3052,3059,3066,3073,3087,3094,3101,3108,3122,3129,3136,3143,3157,3164,3171,3178,3192,3199,


In [3]:
# Using generators and list comprehension
print( *(i for i in range(2000,3201) if i%7==0 and i%5!=0),sep=',')

2002,2009,2016,2023,2037,2044,2051,2058,2072,2079,2086,2093,2107,2114,2121,2128,2142,2149,2156,2163,2177,2184,2191,2198,2212,2219,2226,2233,2247,2254,2261,2268,2282,2289,2296,2303,2317,2324,2331,2338,2352,2359,2366,2373,2387,2394,2401,2408,2422,2429,2436,2443,2457,2464,2471,2478,2492,2499,2506,2513,2527,2534,2541,2548,2562,2569,2576,2583,2597,2604,2611,2618,2632,2639,2646,2653,2667,2674,2681,2688,2702,2709,2716,2723,2737,2744,2751,2758,2772,2779,2786,2793,2807,2814,2821,2828,2842,2849,2856,2863,2877,2884,2891,2898,2912,2919,2926,2933,2947,2954,2961,2968,2982,2989,2996,3003,3017,3024,3031,3038,3052,3059,3066,3073,3087,3094,3101,3108,3122,3129,3136,3143,3157,3164,3171,3178,3192,3199


# Question 2 - factorial

Question:  
Write a program which can compute the factorial of a given numbers.The results should be printed in a comma-separated sequence on a single line. Suppose the following input is supplied to the program: 8 Then, the output should be:40320

Hints:  
In case of input data being supplied to the question, it should be assumed to be a console input.

#### Solution 1:Using While Loop

In [4]:
def factorial(num):
    res = 1
    while num > 1:
        res *= num
        num -= 1
    return res
factorial(9)

362880

#### Solution 2: Using For Loop

In [5]:
def factorial(num):
    res = 1
    for i in range(2,num+1):
        res *= i
    return res
factorial(8)

40320

#### Solution 3: Using Lambda Function / recursion

In [6]:
n = int(input())

def shortFact(x):
    return 1 if x<=1 else x*shortFact(x-1)
print(shortFact(n))

8
40320


# Question 3 - dict
Question:
With a given integral number n, write a program to generate a dictionary that contains (i, i x i) such that is an integral number between 1 and n (both included). and then the program should print the dictionary.Suppose the following input is supplied to the program: 8

Then, the output should be:  
{1: 1, 2: 4, 3: 9, 4: 16, 5: 25, 6: 36, 7: 49, 8: 64}

Hints:  
In case of input data being supplied to the question, it should be assumed to be a console input.Consider use dict()

#### Solution:  for loop

In [7]:
n = int(input())
d = {}
for i in range(1,n+1):
    d[i] = i*i
print(d)

8
{1: 1, 2: 4, 3: 9, 4: 16, 5: 25, 6: 36, 7: 49, 8: 64}


#### Solution: Using dictionary comprehension

In [8]:
n = int(input())
d = {i:i*i for i in range(1,n+1)}
print(d)

8
{1: 1, 2: 4, 3: 9, 4: 16, 5: 25, 6: 36, 7: 49, 8: 64}


# Question 4: list + tuple
Question:  
Write a program which accepts a sequence of comma-separated numbers from console and generate a list and a tuple which contains every number.Suppose the following input is supplied to the program:  34,67,55,33,12,98

Then, the output should be: ['34', '67', '55', '33', '12', '98'] ('34', '67', '55', '33', '12', '98')

Hints:  
In case of input data being supplied to the question, it should be assumed to be a console input.tuple() method can convert list to tuple

In [9]:
lst = input().split(',')
tpl = tuple(lst)
print(lst,tpl)

34,67,55,33,12,98
['34', '67', '55', '33', '12', '98'] ('34', '67', '55', '33', '12', '98')


# Question 5
Question: Define a class which has at least two methods:  
1. getString: to get a string from console input  
2. printString: to print the string in upper case.  
Also please include simple test function to test the class methods.

Hints: _Use init method to construct some parameters_

In [10]:
class IOString:
    def _init_(self):
        pass
    def getString(self):
        self.s = input()
    def printString(self):
        print(self.s.upper())
        
yy = IOString()
yy.getString()
yy.printString()

IOString("ds")
IOSTRING("DS")


# Question 6
Question: Write a program that calculates and prints the value according to the given formula:  
Q = Square root of [(2 C D)/H]

Following are the fixed values of C and H:  
C is 50. H is 30.  
D is the variable whose values should be input to your program in a comma-separated sequence.   
For example Let us assume the following comma separated input sequence is given to the program: 100,150,180  
The output of the program should be: 18,22,24

Hints:
If the output received is in decimal form, it should be rounded off to its nearest value (for example, if the output received is 26.0, it should be printed as 26).In case of input data being supplied to the question, it should be assumed to be a console input.

In [11]:
from math import sqrt
C, H = 50, 30

def calculate(D):
    return sqrt((2*C*D)/H)

D = [int(i) for i in input().split(',')]
D = [calculate(i) for i in D]
D = [round(i) for i in D]
D = [str(i) for i in D]
print(','.join(D))


100,150,180
18,22,24


In [12]:
from math import sqrt
C, H = 50, 30

def calculate(D):
    return sqrt((2*C*D)/H)

D = [str(round(calculate(int(i)))) for i in input().split(',')]
print(','.join(D))

100,150,180
18,22,24


In [13]:
from math import sqrt
C, H = 50, 30

def calculate(D):
    D = int(D)
    return str(int(sqrt((2*C*D)/H)))

D = input().split(',')
D = list(map(calculate,D))
print(",".join(D))

100,150,180
18,22,24


# Question 7
Question: Write a program which takes 2 digits, X,Y as input and generates a 2-dimensional array. The element value in the i-th row and j-th column of the array should be i * j.

Note:   
i=0,1.., X-1; j=0,1 ..,Y-1. Suppose the following inputs are given to the program: 3,5  
Then, the output of the program should be:
[[0, 0, 0, 0, 0], [0, 1, 2, 3, 4], [0, 2, 4, 6, 8]]

Hints:
Note: In case of input data being supplied to the question, it should be assumed to be a console input in a comma-separated form.

In [14]:
x, y = map(int,input().split(','))
lst = []
for i in range(x):
    tmp = []
    for j in range(y):
        tmp.append(i*j)
    lst.append(tmp)
print(lst)

3,5
[[0, 0, 0, 0, 0], [0, 1, 2, 3, 4], [0, 2, 4, 6, 8]]


In [15]:
x, y = map(int,input().split(','))
res = [[i*j for j in range(y)] for i in range(x)]
print(res)

3,5
[[0, 0, 0, 0, 0], [0, 1, 2, 3, 4], [0, 2, 4, 6, 8]]


# Question 8
Question:  
Write a program that accepts a comma separated sequence of words as input and prints the words in a comma-separated sequence after sorting them alphabetically.

Suppose the following input is supplied to the program: without,hello,bag,world    
Then, the output should be: bag,hello,without,world  

Hints:
In case of input data being supplied to the question, it should be assumed to be a console input.

In [16]:
words = input().split(',')
words.sort()
print(','.join(words))

without,hello,bag,world
bag,hello,without,world


# Question 9
Question: Write a program that accepts sequence of lines as input and prints the lines after making all characters in the sentence capitalized.
Suppose the following input is supplied to the program:   
Hello world      
Practice makes perfect  

Then, the output should be:  
HELLO WORLD  
PRACTICE MAKES PERFECT  

Hints:  
In case of input data being supplied to the question, it should be assumed to be a console input.



In [17]:
lst = []
while True:
    x = input()
    if len(x) == 0:
        break
    lst.append(x.upper())
for word in lst:
    print(word)

Hello world Practice makes perfect

HELLO WORLD PRACTICE MAKES PERFECT


# Question 10
Question:  
Write a program that accepts a sequence of whitespace separated words as input and prints the words after removing all duplicate words and sorting them alphanumerically.

Suppose the following input is supplied to the program:  
hello world and practice makes perfect and hello world again  
Then, the output should be:  
again and hello makes perfect practice world

Hints:  
In case of input data being supplied to the question, it should be assumed to be a console input.We use set container to remove duplicated data automatically and then use sorted() to sort the data.

In [18]:
words = input().split()
for word in words:
    if(words.count(word)>1):
        words.remove(word)
words.sort()
print(' '.join(words))

hello world and practice makes perfect and hello world again
again and hello makes perfect practice world


# Question 11
Question  
Write a program which accepts a sequence of comma separated 4 digit binary numbers as its input and then check whether they are divisible by 5 or not. The numbers that are divisible by 5 are to be printed in a comma separated sequence.

Example: 0100,0011,1010,1001

Then the output should be: 1010

Notes: Assume the data is input by console.  
Hints: In case of input data being supplied to the question, it should be assumed to be a console input.

In [19]:
digits = input().split(',')
for digit in digits:
    if int(digit,2) % 5 == 0:
        print(digit)

 0100,0011,1010,1001
1010


In [20]:
digits = input().split(',')
def check(x):
    return int(x,2) % 5 == 0

data = list(filter(check,digits))
print(','.join(data))

 0100,0011,1010,1001
1010


In [21]:
digits = input().split(',')
digits = list(filter(lambda x:int(x,2) % 5 == 0,digits))
print(','.join(digits))

 0100,0011,1010,1001
1010


# Question 12
Question:
Write a program, which will find all such numbers between 1000 and 3000 (both included) such that each digit of the number is an even number.The numbers obtained should be printed in a comma-separated sequence on a single line.

Hints:
In case of input data being supplied to the question, it should be assumed to be a console input.

In [22]:
def check(element):
    return all(
        ord(i) % 2 == 0 for i in element)  # all returns True if all digits i is even in element


lst = [str(i) for i in range(1000, 3001)]  # creates list of all given numbers with string data type
lst = list(filter(check, lst))  # filter removes element from list if check condition fails
print(",".join(lst))

2000,2002,2004,2006,2008,2020,2022,2024,2026,2028,2040,2042,2044,2046,2048,2060,2062,2064,2066,2068,2080,2082,2084,2086,2088,2200,2202,2204,2206,2208,2220,2222,2224,2226,2228,2240,2242,2244,2246,2248,2260,2262,2264,2266,2268,2280,2282,2284,2286,2288,2400,2402,2404,2406,2408,2420,2422,2424,2426,2428,2440,2442,2444,2446,2448,2460,2462,2464,2466,2468,2480,2482,2484,2486,2488,2600,2602,2604,2606,2608,2620,2622,2624,2626,2628,2640,2642,2644,2646,2648,2660,2662,2664,2666,2668,2680,2682,2684,2686,2688,2800,2802,2804,2806,2808,2820,2822,2824,2826,2828,2840,2842,2844,2846,2848,2860,2862,2864,2866,2868,2880,2882,2884,2886,2888


In [23]:
lst = [str(i) for i in range(1000,3001)]
lst = list(filter(lambda i: all(ord(j)%2==1 for j in i),lst))
print(','.join(lst))

1111,1113,1115,1117,1119,1131,1133,1135,1137,1139,1151,1153,1155,1157,1159,1171,1173,1175,1177,1179,1191,1193,1195,1197,1199,1311,1313,1315,1317,1319,1331,1333,1335,1337,1339,1351,1353,1355,1357,1359,1371,1373,1375,1377,1379,1391,1393,1395,1397,1399,1511,1513,1515,1517,1519,1531,1533,1535,1537,1539,1551,1553,1555,1557,1559,1571,1573,1575,1577,1579,1591,1593,1595,1597,1599,1711,1713,1715,1717,1719,1731,1733,1735,1737,1739,1751,1753,1755,1757,1759,1771,1773,1775,1777,1779,1791,1793,1795,1797,1799,1911,1913,1915,1917,1919,1931,1933,1935,1937,1939,1951,1953,1955,1957,1959,1971,1973,1975,1977,1979,1991,1993,1995,1997,1999


# Question 13
Question:
Write a program that accepts a sentence and calculate the number of letters and digits.  
Suppose the following input is supplied to the program:  
hello world! 123   
Then, the output should be:  
LETTERS 10  
DIGITS 3    
Hints: In case of input data being supplied to the question, it should be assumed to be a console input.

In [24]:
words = input()
letters, digits = 0, 0
for i in words:
    if i.isalpha():
        letters += 1
    if i.isdigit():
        digits += 1
print('LETTERS ',letters)
print('DIGITS ',digits)
print("==============")
print("LETTERS {0}\nDIGITS {1}".format(letters, digits))

hello world! 123
LETTERS  10
DIGITS  3
LETTERS 10
DIGITS 3


# Question 14
Question:
Write a program that accepts a sentence and calculate the number of upper case letters and lower case letters.

Suppose the following input is supplied to the program: Hello world!  
Then, the output should be: 
UPPER CASE 1  
LOWER CASE 9  

Hints:
In case of input data being supplied to the question, it should be assumed to be a console input.

In [26]:
sentence = input()
upperCase, lowerCase = 0, 0
for i in sentence:
    if i.isupper():
        upperCase += 1
    elif i.islower():
        lowerCase += 1
print("UPPER CASE {0}\nLOWER CASE {1}".format(upperCase,lowerCase))

 Hello world!
UPPER CASE 1
LOWER CASE 9


In [27]:
sentence = input()
upperCase = sum(i.isupper() for i in sentence)
lowerCase = sum(i.islower() for i in sentence)
print("UPPER CASE {0}\nLOWER CASE {1}".format(upperCase,lowerCase))

Hello world!
UPPER CASE 1
LOWER CASE 9


# Question 15
Question:
Write a program that computes the value of a+aa+aaa+aaaa with a given digit as the value of a.

Suppose the following input is supplied to the program: 9  
Then, the output should be: 11106

Hints: In case of input data being supplied to the question, it should be assumed to be a console input.



In [36]:
num = input()
res = 0
string = ''
for i in range(4):
    string += str(num)
    res += int(string)
print(res)

9
11106


# Question 16
Question:
Use a list comprehension to square each odd number in a list. The list is input by a sequence of comma-separated numbers.   
Suppose the following input is supplied to the program: 1,2,3,4,5,6,7,8,9

Then, the output should be: 1,9,25,49,81

Hints: In case of input data being supplied to the question, it should be assumed to be a console input.

In [47]:
numbers = list(str(int(num)**2) for num in input().split(',') if int(num)%2==1 )
print(','.join(numbers))

1,2,3,4,5,6,7,8,9
1,9,25,49,81


# Question 17
Question:
Write a program that computes the net amount of a bank account based a transaction log from console input. The transaction log format is shown as following:
 
D 100  
W 200  
D means deposit while W means withdrawal.  

Suppose the following input is supplied to the program:  
D 300  
D 300  
W 200  
D 100  
Then, the output should be: 500

In [51]:
total = 0
while True:
    s = input().split()
    if not s:
        break
    op, num = map(str,s)
    if op == 'D':
        total += int(num)
    elif op == 'W':
        total -= int(num)
print(num)

D 300
D 300
W 200
D 100

100


In [56]:
balance = 0
while True:
    action = input("Deposit/Withdrow/Balance/Quit? D/W/B/Q:").lower()
    if action == 'd':
        deposit = input("How much would you like to deposit?")
        balance += int(deposit)
    elif action == 'w':
        withdrow = input("How much would you like to withdrow?")
        balance -= int(withdrow)
    elif action == 'b':
        print('balance:{0}'.format(balance))
    elif action == 'q':
        break       

Deposit/Withdrow/Balance/Quit? D/W/B/Q:d
How much would you like to deposit?300
Deposit/Withdrow/Balance/Quit? D/W/B/Q:d
How much would you like to deposit?300
Deposit/Withdrow/Balance/Quit? D/W/B/Q:w
How much would you like to withdrow?200
Deposit/Withdrow/Balance/Quit? D/W/B/Q:d
How much would you like to deposit?100
Deposit/Withdrow/Balance/Quit? D/W/B/Q:b
balance:500
Deposit/Withdrow/Balance/Quit? D/W/B/Q:q


# Question 18
Question:
A website requires the users to input username and password to register. Write a program to check the validity of password input by users.  
Following are the criteria for checking the password:  
At least 1 letter between [a-z]  
At least 1 number between [0-9]  
At least 1 letter between [A-Z]  
At least 1 character from [$#@]  
Minimum length of transaction password: 6  
Maximum length of transaction password: 12  
Your program should accept a sequence of comma separated passwords and will check them according to the above criteria. Passwords that match the criteria are to be printed, each separated by a comma.

Example  
If the following passwords are given as input to the program:  
ABd1234@1,a F1#,2w3E*,2We3345    
Then, the output of the program should be:    
ABd1234@1

Hints: In case of input data being supplied to the question, it should be assumed to be a console input.

In [64]:
def qualifiedPassword(x):
    cnt = (len(x) >= 6 and len(x) <= 12)
    for i in x:
        if i.isupper():
            cnt += 1
            break
    for i in x:
        if i.islower():
            cnt += 1
            break
    for i in x:
        if i.isnumeric():
            cnt += 1
            break 
    for i in x:
        if i == "@" or i == "#" or i == "$":
            cnt += 1
            break
    return cnt == 5

s = input().split(',')
lst = filter(qualifiedPassword,s)
print(','.join(lst))

ABd1234@1,a F1#,2w3E*,2We3345
ABd1234@1


In [63]:
import re
pwds = input().split(',')
lst = []
for pwd in pwds:
    cnt = 0
    cnt += (len(pwd) >= 6 and len(pwd) <= 12)
    cnt += bool(re.search("[a-z]", pwd))
    cnt += bool(re.search("[0-9]", pwd))
    cnt += bool(re.search("[A-Z]", pwd))
    cnt += bool(re.search("[$#@]", pwd))
    if cnt == 5:
        lst.append(pwd)
print(','.join(lst))

ABd1234@1,a F1#,2w3E*,2We3345
ABd1234@1


# Question 19  
Question:
You are required to write a program to sort the (name, age, score) tuples by ascending order where name is string, age and score are numbers. The tuples are input by console. The sort criteria is:

1: Sort based on name  
2: Then sort based on age  
3: Then sort by score  
The priority is that name > age > score.

If the following tuples are given as input to the program:

Tom,19,80

John,20,90

Jony,17,91

Jony,17,93

Json,21,85

Then, the output of the program should be:

[('John', '20', '90'), ('Jony', '17', '91'), ('Jony', '17', '93'), ('Json', '21', '85'), ('Tom', '19', '80')]

Hints:
In case of input data being supplied to the question, it should be assumed to be a console input.We use itemgetter to enable multiple sort keys.

In [2]:
lst = []
while True:
    s = input().split(',')
    if not s[0]:
        break
    lst.append(tuple(s))

lst.sort(key=lambda x: (x[0],x[1],x[2]))
print(lst)

Tom,19,80
John,20,90
Jony,17,91
Jony,17,93
Json,21,85

[('John', '20', '90'), ('Jony', '17', '91'), ('Jony', '17', '93'), ('Json', '21', '85'), ('Tom', '19', '80')]


# Question 20
Question:  
Define a class with a generator which can iterate the numbers, which are divisible by 7, between a given range 0 and n.

Hints:
Consider use class, function and comprehension.



In [12]:
class Divisible:
    def by_seven(self, n):
        for number in range(n + 1):
            if number % 7 == 0:
                yield number


generator = Divisible().by_seven(int(input("Please insert a number. --> ")))
for number in generator:
    print(number)

Please insert a number. --> 54
0
7
14
21
28
35
42
49


# Question 21
Question:
A robot moves in a plane starting from the original point (0,0). The robot can move toward UP, DOWN, LEFT and RIGHT with a given steps. The trace of robot movement is shown as the following:  
UP 5  
DOWN 3  
LEFT 3   
RIGHT 2  

The numbers after the direction are steps. Please write a program to compute the distance from current position after a sequence of movement and original point. If the distance is a float, then just print the nearest integer. 
Example: If the following tuples are given as input to the program:  
UP 5  
DOWN 3  
LEFT 3  
RIGHT 2  
Then, the output of the program should be: 2

Hints: In case of input data being supplied to the question, it should be assumed to be a console input.Here distance indicates to euclidean distance.Import math module to use sqrt function.

In [7]:
from math import sqrt
x, y = 0, 0
while True:
    s = input().split()
    if not s:
        break
    if s[0] == 'UP':
        y += int(s[1])
    elif s[0] == 'DOWN':
        y -= int(s[1])
    elif s[0] == 'LEFT':
        x -= int(s[1])
    elif s[0] == 'RIGHT':
        x += int(s[1])
print(int(sqrt(x**x + y**y)))        

UP 5
DOWN 3
LEFT 3
RIGHT 2

1.7320508075688772


In [3]:
from math import sqrt
x, y = 0, 0
while True:
    dirt = input("Please enter the direction: UP/DOWN/LEFT/RIGHT/QUIT,u/d/l/r/q: ")
    if dirt == 'q':
        break
    if dirt == 'u':
        step = int(input("Please enter the number of steps:"))
        y += step
    if dirt == 'd':
        step = int(input("Please enter the number of steps:"))
        y -= step
    if dirt == 'l':
        step = int(input("Please enter the number of steps:"))
        x -= step
    if dirt == 'r':
        step = int(input("Please enter the number of steps: "))
        x += step
    print(int(sqrt(x*x+y*y)))

Please enter the direction: UP/DOWN/LEFT/RIGHT/QUIT,u/d/l/r/q: u
Please enter the number of steps:5
5
Please enter the direction: UP/DOWN/LEFT/RIGHT/QUIT,u/d/l/r/q: d
Please enter the number of steps:3
2
Please enter the direction: UP/DOWN/LEFT/RIGHT/QUIT,u/d/l/r/q: l
Please enter the number of steps:3
3
Please enter the direction: UP/DOWN/LEFT/RIGHT/QUIT,u/d/l/r/q: r
Please enter the number of steps: 2
2
Please enter the direction: UP/DOWN/LEFT/RIGHT/QUIT,u/d/l/r/q: 1
2
Please enter the direction: UP/DOWN/LEFT/RIGHT/QUIT,u/d/l/r/q: q


# Question 22  
Question:  
Write a program to compute the frequency of the words from the input. The output should output after sorting the key alphanumerically.  
Suppose the following input is supplied to the program:  
New to Python or choosing between Python 2 and Python 3? Read Python 2 or Python 3.  
Then, the output should be:  
2:2  
3.:1  
3?:1  
New:1  
Python:5  
Read:1  
and:1  
between:1  
choosing:1  
or:2  
to:1  
Hints:In case of input data being supplied to the question, it should be assumed to be a console input.

In [2]:
ss = input().split()
words = sorted(set(ss))
for i in words:
    print("{0}:[1]".format(i,ss.count(i)))

New to Python or choosing between Python 2 and Python 3? Read Python 2 or Python 3.
2:[1]
3.:[1]
3?:[1]
New:[1]
Python:[1]
Read:[1]
and:[1]
between:[1]
choosing:[1]
or:[1]
to:[1]


In [5]:
ss = input().split()
import collections
d = collections.Counter(ss)
d = sorted(d.items())
print(d)
for items in d:
    print("{0}:{1}".format(items[0],items[1]))

New to Python or choosing between Python 2 and Python 3? Read Python 2 or Python 3.
[('2', 2), ('3.', 1), ('3?', 1), ('New', 1), ('Python', 5), ('Read', 1), ('and', 1), ('between', 1), ('choosing', 1), ('or', 2), ('to', 1)]
2:2
3.:1
3?:1
New:1
Python:5
Read:1
and:1
between:1
choosing:1
or:2
to:1


In [6]:
from pprint import pprint

p = input().split()
pprint({i: p.count(i) for i in p})

New to Python or choosing between Python 2 and Python 3? Read Python 2 or Python 3.
{'2': 2,
 '3.': 1,
 '3?': 1,
 'New': 1,
 'Python': 5,
 'Read': 1,
 'and': 1,
 'between': 1,
 'choosing': 1,
 'or': 2,
 'to': 1}


# Question 23
Question:
Write a method which can calculate square value of number  
Hints:
Using the operator which can be written as n**p where means n^p

In [7]:
num = int(input())
print(num**2)

3
9


# Question 24
Question:
Python has many built-in functions, and if you do not know how to use it, you can read document online or find some books. But Python has a built-in document function for every built-in functions.

_Please write a program to print some Python built-in functions documents, such as abs(), int(), rawinput()

And add document for your own function

Hints:
The built-in document method is doc

In [9]:
print(str.__doc__)

str(object='') -> str
str(bytes_or_buffer[, encoding[, errors]]) -> str

Create a new string object from the given object. If encoding or
errors is specified, then the object must expose a data buffer
that will be decoded using the given encoding and error handler.
Otherwise, returns the result of object.__str__() (if defined)
or repr(object).
encoding defaults to sys.getdefaultencoding().
errors defaults to 'strict'.


In [10]:
print(sorted.__doc__)

Return a new list containing all items from the iterable in ascending order.

A custom key function can be supplied to customize the sort order, and the
reverse flag can be set to request the result in descending order.


In [11]:
def pow(n,p):
    """
    param n: This is any integer number
    param p: This is power over n
    return:  n to the power p = n^p
    """
    return n ** p
print(pow.__doc__)


    param n: This is any integer number
    param p: This is power over n
    return:  n to the power p = n^p
    


# Question 25
Question:  
Define a class, which have a class parameter and have a same instance parameter.

Hints:  
Define an instance parameter, need add it in init method.You can init an object with construct parameter or set the value later

In [18]:
class Car:
    name = "Car"

    def __init__(self, name=None):
        self.name = name


honda = Car("Honda")
print(f"{Car.name} is {honda.name}")

Car is Honda


In [19]:
toyota = Car()
toyota.name = "Toyota"
print(f"{Car.name} is {toyota.name}")

Car is Toyota


# Question 26
Question:
Define a function which can compute the sum of two numbers.

Hints:
Define a function with two numbers as arguments. You can compute the sum in the function and return the value.

In [22]:
def Sum(a:int,b:int) ->int:
    return a+b
Sum(1,3)

4

In [23]:
sum = lambda n1,n2:n1+n2
sum(1,3)

4

# Question 27

Question:
Define a function that can convert a integer into a string and print it in console.  
Hints:
Use str() to convert a number to string.

In [25]:
def convIntToStr(num:int) -> str:
    return str(num)
convIntToStr(3343)

'3343'

In [27]:
conv = lambda num: str(num)
print(conv(23))
print(type(conv(23)))

23
<class 'str'>


# Question 28
Question:
Define a function that can receive two integer numbers in string form and compute their sum and then print it in console.  
Hints:
Use int() to convert a string to integer.

In [28]:
def sumIntfromStr(str1:str,str2:str) -> int:
    return sum(int(str1),int(str2))
sumIntfromStr('23','32')

55

In [ ]:
sumStr = lambda str1,str2: int(str1)+int(str2)

# Question 29 - str+str
Question:
Define a function that can accept two strings as input and concatenate them and then print it in console.

Hints:
Use + sign to concatenate the strings.

In [29]:
def concatenateStr(str1:str,str2:str)->str:
    return str1 + str2
concatenateStr("dr3","fr3")

'dr3fr3'

In [ ]:
concatenateStr2 = lambda str1,str2: str1 + str2

# Question 30 - str, max(a,b,key=len)
Question:
Define a function that can accept two strings as input and print the string with maximum length in console. If two strings have the same length, then the function should print all strings line by line.

Hints:
Use len() function to get the length of a string.

In [31]:
def printMax(str1,str2)->str:
    l1,l2 = len(str1),len(str2)
    if l1 < l2:
        print(str2)
    elif l1 > l2:
        print(str1)
    else:
        print(str1)
        print(str2)
printMax("sds","fd2")

sds
fd2


In [35]:
printMaxStr = lambda s1,s2: print(max(s1,s2,key=len)) if len(s1)!= len(s2) else print(s1+'\n'+s2)
printMax("sds","fd2")

sds
fd2


# Question 31 - dict
Question:
Define a function which can print a dictionary where the keys are numbers between 1 and 20 (both included) and the values are square of keys.

Hints:
Use dict[key]=value pattern to put entry into a dictionary.Use ** operator to get power of a number.Use range() for loops.

In [37]:
def printDict(n):
    d = {}
    for i in range(1,n+1):
        d[i] = i*i
    print(d)
printDict(20)

{1: 1, 2: 4, 3: 9, 4: 16, 5: 25, 6: 36, 7: 49, 8: 64, 9: 81, 10: 100, 11: 121, 12: 144, 13: 169, 14: 196, 15: 225, 16: 256, 17: 289, 18: 324, 19: 361, 20: 400}


In [40]:
def printDict():
    dict = {i:i*i for i in range(1,21)}
    print(dict)
printDict()

{1: 1, 2: 4, 3: 9, 4: 16, 5: 25, 6: 36, 7: 49, 8: 64, 9: 81, 10: 100, 11: 121, 12: 144, 13: 169, 14: 196, 15: 225, 16: 256, 17: 289, 18: 324, 19: 361, 20: 400}


# Question 32 - dict, keys(), items()
Question:
Define a function which can generate a dictionary where the keys are numbers between 1 and 20 (both included) and the values are square of keys. The function should just print the keys only.

Hints:
Use dict[key]=value pattern to put entry into a dictionary.Use ** operator to get power of a number.Use range() for loops.Use keys() to iterate keys in the dictionary. Also we can use item() to get key/value pairs.

In [44]:
def generateDict():
    d = {i:i**2 for i in range(1,21)}
    print(d.keys())
    print(d.items())
generateDict()

dict_keys([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20])
dict_items([(1, 1), (2, 4), (3, 9), (4, 16), (5, 25), (6, 36), (7, 49), (8, 64), (9, 81), (10, 100), (11, 121), (12, 144), (13, 169), (14, 196), (15, 225), (16, 256), (17, 289), (18, 324), (19, 361), (20, 400)])


# Question 33
Question:
Define a function which can generate and print a list where the values are square of numbers between 1 and 20 (both included).

Hints:
Use ** operator to get power of a number.Use range() for loops.Use list.append() to add values into a list.

In [45]:
def generateAndPrint():
    lst = [i**2 for i in range(1,21)]
    print(lst)
generateAndPrint()    

[1, 4, 9, 16, 25, 36, 49, 64, 81, 100, 121, 144, 169, 196, 225, 256, 289, 324, 361, 400]


In [46]:
def generateAndPrint2():
    lst = []
    for i in range(1,21):
        lst.append(i**2)
    print(lst)
generateAndPrint2()

[1, 4, 9, 16, 25, 36, 49, 64, 81, 100, 121, 144, 169, 196, 225, 256, 289, 324, 361, 400]


# Question 34
Question:
Define a function which can generate a list where the values are square of numbers between 1 and 20 (both included). Then the function needs to print the first 5 elements in the list.

Hints:
Use ** operator to get power of a number.Use range() for loops.Use list.append() to add values into a list.Use [n1:n2] to slice a list

In [47]:
def generateAndPrint5():
    lst = []
    for i in range(1,21):
        lst.append(i**2)
    print(lst[:5])
generateAndPrint5()    

[1, 4, 9, 16, 25]


# Question 35
Question:
Define a function which can generate a list where the values are square of numbers between 1 and 20 (both included). Then the function needs to print the last 5 elements in the list.

Hints:
Use ** operator to get power of a number.Use range() for loops.Use list.append() to add values into a list.Use [n1:n2] to slice a list

In [48]:
def generateLst():
    lst = [i**2 for i in range(1,21)]
    print(lst[-5:])
generateLst()

[256, 289, 324, 361, 400]


In [52]:
generateLst2 = lambda:([i**2 for i in range(1,21)][-5:])
print(*generateLst2(),sep='\n')

256
289
324
361
400


# Question 36
Question:
Define a function which can generate a list where the values are square of numbers between 1 and 20 (both included). Then the function needs to print all values except the first 5 elements in the list.

Hints: Use ** operator to get power of a number.Use range() for loops.Use list.append() to add values into a list.Use [n1:n2] to slice a list

In [54]:
generateLst2 = lambda:([i**2 for i in range(1,21)][5:])
print(*generateLst2(),sep='\n')

36
49
64
81
100
121
144
169
196
225
256
289
324
361
400


# Question 37
Question:
Define a function which can generate and print a tuple where the value are square of numbers between 1 and 20 (both included).

Hints:
Use ** operator to get power of a number.Use range() for loops.Use list.append() to add values into a list.Use tuple() to get a tuple from a list.

In [59]:
def printTup():
    lst = [i**2 for i in range(1,21)]
    print(tuple(lst))
printTup()

(1, 4, 9, 16, 25, 36, 49, 64, 81, 100, 121, 144, 169, 196, 225, 256, 289, 324, 361, 400)


# Day 11
# Question 38 - tuple
Question:
With a given tuple (1,2,3,4,5,6,7,8,9,10), write a program to print the first half values in one line and the last half values in one line.

Hints:
Use [n1:n2] notation to get a slice from a tuple.

In [60]:
def printTup(tup):
    print(tup[:len(tup)//2])
    print(tup[len(tup)//2:])
tup = (1,2,3,4,5,6,7,8,9,10)
printTup(tup)

(1, 2, 3, 4, 5)
(6, 7, 8, 9, 10)


# Question 39
Question:
Write a program to generate and print another tuple whose values are even numbers in the given tuple (1,2,3,4,5,6,7,8,9,10).

Hints:
Use "for" to iterate the tuple. Use tuple() to generate a tuple from a list.

In [64]:
def evenEle(tup):
    lst = []
    for t in tup:
        if t % 2 == 0:
            lst.append(t)
    print(tuple(lst))
tup = (1,2,3,4,5,6,7,8,9,10)
evenEle(tup)

(2, 4, 6, 8, 10)


In [65]:
tup = (1,2,3,4,5,6,7,8,9,10)
res = tuple(i for i in tup if i%2==0)
print(res)

(2, 4, 6, 8, 10)


In [67]:
tup = (1,2,3,4,5,6,7,8,9,10)
res = tuple(filter(lambda x: x%2==0,tup))
print(res)

(2, 4, 6, 8, 10)


# Question 40
Question:
Write a program which accepts a string as input to print "Yes" if the string is "yes" or "YES" or "Yes", otherwise print "No".

Hints:
Use if statement to judge condition.

In [70]:
string = input("Please enter string: ")
if string == "yes" or string == "YES" or string == "Yes":
    print("Yes")
else:
    print("No")

Please enter string: YEs
No


# Question 41 - map()
Question:
Write a program which can map() to make a list whose elements are square of elements in [1,2,3,4,5,6,7,8,9,10].

Hints:
Use map() to generate a list.Use lambda to define anonymous functions.

In [74]:
lst = [1,2,3,4,5,6,7,8,9,10]
res = map(lambda i:i**2,lst)
print(list(res))

[1, 4, 9, 16, 25, 36, 49, 64, 81, 100]


# Question 42
Question:
Write a program which can map() and filter() to make a list whose elements are square of even number in [1,2,3,4,5,6,7,8,9,10].

Hints:
Use map() to generate a list.Use filter() to filter elements of a list.Use lambda to define anonymous functions.

In [78]:
lst = [1,2,3,4,5,6,7,8,9,10]
res = map(lambda i:i**2,filter(lambda x:x%2==0,lst))
print(list(res))

[4, 16, 36, 64, 100]


In [79]:
def even(x):
    return x % 2 == 0

def squer(x):
    return x * x

li = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
li = map(
    squer, filter(even, li)
)  # first filters number by even number and the apply map() on the resultant elements
print(list(li))

[4, 16, 36, 64, 100]


# Question 43
Question:
Write a program which can filter() to make a list whose elements are even number between 1 and 20 (both included).

Hints:
Use filter() to filter elements of a list.Use lambda to define anonymous functions.

In [82]:
lst = [i for i in range(1,21)]
res = filter(lambda x: x%2==0,lst)
print(list(res))

[2, 4, 6, 8, 10, 12, 14, 16, 18, 20]


In [83]:
def even(x):
    return x % 2 == 0
print(list(filter(even,range(1,21))))

[2, 4, 6, 8, 10, 12, 14, 16, 18, 20]


# Question 44
Question:
Write a program which can map() to make a list whose elements are square of numbers between 1 and 20 (both included).

Hints:
Use map() to generate a list. Use lambda to define anonymous functions.

In [84]:
print(list(map(lambda x:x**2,range(1,21))))

[1, 4, 9, 16, 25, 36, 49, 64, 81, 100, 121, 144, 169, 196, 225, 256, 289, 324, 361, 400]


# Question 45
Question:
Define a class named American which has a static method called printNationality.

Hints:
Use @staticmethod decorator to define class static method.There are also two more methods.To know more, go to this link. https://realpython.com/instance-class-and-static-methods-demystified/

In [85]:
class American:
    @staticmethod 
    def printNationality():
        print("I am Ameriaca")
American.printNationality()

I am Ameriaca


# Question 46
Question:
Define a class named American and its subclass NewYorker.

Hints:
Use class Subclass(ParentClass) to define a subclass.

In [86]:
class American:
    pass 
class NewYorker(American):
    pass

america = American()
newyorker =  NewYorker()
print(america)
print(newyorker)

# Question 47
Question
Define a class named Circle which can be constructed by a radius. The Circle class has a method which can compute the area.

Hints
Use def methodName(self) to define a method.

In [90]:
import math
class Circle:
    def __init__(self,r):
        self.r = r
    def area(self):
        return math.pi*(self.r**2)
print(Circle(2).area())

12.566370614359172


# Question 48
Question
Define a class named Rectangle which can be constructed by a length and width. The Rectangle class has a method which can compute the area.

Hints
Use def methodName(self) to define a method.

In [93]:
class Rectangle:
    def __init__(self,l,w):
        self.length = l
        self.width = w
    def area(self):
        return self.length * self.width

print(Rectangle(2,3).area())

6


# Question 49
Question
Define a class named Shape and its subclass Square. The Square class has an init function which takes a length as argument. Both classes have a area function which can print the area of the shape where Shape's area is 0 by default.

Hints
To override a method in super class, we can define a method with the same name in the super class.



In [94]:
class Shape:
    def __init__(self):
        pass
    def area(self):
        return 0

class Square(Shape):
    def __init__(self, length=0):
        Shape.__init__(self)
        self.length = length
    '''
    def __init__(self,l):
        self.length = l
    '''
    def area(self):
        return self.length**2

print(Square(2).area())

4


# Question 50 - raise 异常
Question
Please raise a RuntimeError exception.

Hints
Use raise to raise an exception.



In [102]:
raise RuntimeError("something wrong")

RuntimeError: something wrong

# Question 51 - try/except/else
Question
Write a function to compute 5/0 and use try/except to catch the exceptions.

Hints
Use try/except to catch exceptions.

In [1]:
def divide():
    return 5 / 0
try:
    divide()
except ZeroDivisionError:
    print("Why on earth you are dividing a number by ZERO!!")
except:
    print("other exception")  

Why on earth you are dividing a number by ZERO!!


# Question 52 - define an exception class
Question
Define a custom exception class which takes a string message as attribute.

Hints
To define a custom exception, we need to define a class inherited from Exception.



In [4]:
class CustomerException(Exception):
    """Exception raised for custom purpose

    Attributes:
        message -- explanation of the error
    """
    def __init__(self,message):
        self.message = message

num = int(input()) 
try:
    # 正常执行代码块
    if num < 10:
        raise CustomerException("num < 10") #抛出异常 < 10
    elif num > 10:
        raise CustomerException("num > 10") #抛出异常 > 10
        
#发生异常时，执行except下代码块
except CustomerException as ce:
    print("The error raised: ",ce.message)
    
# else: 不发生异常执行该代码块

20
The error raised:  num > 10


# Question 53 - re/regularization findall \w+
Question:
Assuming that we have some email addresses in the "username@companyname.com" format, please write program to print the user name of a given email address. Both user names and company names are composed of letters only.

Example: If the following email address is given as input to the program: john@google.com

Then, the output of the program should be: john

In case of input data being supplied to the question, it should be assumed to be a console input.

Hints:
Use \w to match letters

In [8]:
emailAdress = input()
name = emailAdress.split('@')[0]
print(name)
#username@companyname.com

john@google.com
john


In [9]:
import re

email = "john@google.com elise@python.com"
pattern = "(\w+)@\w+.com"
ans = re.findall(pattern, email)
print(ans)

['john', 'elise']


# Question 54
Question
Assuming that we have some email addresses in the "username@companyname.com" format, please write program to print the company name of a given email address. Both user names and company names are composed of letters only.

Example: If the following email address is given as input to the program: john@google.com

Then, the output of the program should be: google

In case of input data being supplied to the question, it should be assumed to be a console input.

Hints
Use \w to match letters.

In [13]:
import re
email = "john@google.com"
pattern = "\w+@(\w+).com"
ans = re.findall(pattern,email)
print(ans)

['google']


# Question 55
Question
Write a program which accepts a sequence of words separated by whitespace as input to print the words composed of digits only.

Example: If the following words is given as input to the program: 2 cats and 3 dogs.

Then, the output of the program should be: ['2', '3']

In case of inpu^mt data being supplied to the question, it should be assumed to be a console input.

Hints
Use re.findall() to find all substring using regex.

In [16]:
ans = re.findall("[0-9]","2 cats and 3 dogs.")
print(ans)

['2', '3']


In [17]:
import re
words = "2 cats and 3 dogs."
pattern = "\d+"
ans = re.findall(pattern,words)
print(ans)

['2', '3']


# Question 56
Question
Print a unicode string "hello world".

Hints
Use u'strings' format to define unicode string.

In [19]:
unicodeString = u"hello world!"
print(unicodeString)

hello world!


# Question 57
Question:
Write a program to read an ASCII string and to convert it to a unicode string encoded by utf-8.

Hints:
Use unicode()/encode() function to convert.



In [21]:
words = "Write a program"
u = words.encode("utf-8")
print(u)

b'Write a program'


# Question 58
Question: Write a special comment to indicate a Python source code file is in unicode.

Hints: Use unicode() function to convert.



In [22]:
# -*- coding: utf-8 -*-

In [28]:
print('\u0394')
print('\U00000394') #\U + 8位16进制数
print('\N{GREEK CAPITAL LETTER DELTA}')  #\N{字符名称}

Δ
Δ
Δ


# Question 59
Question: Write a program to compute 1/2+2/3+3/4+...+n/n+1 with a given n input by console (n>0). 

Example: If the following n is given as input to the program:5  

Then, the output of the program should be: 3.55  

In case of input data being supplied to the question, it should be assumed to be a console input.  

Hints: Use float() to convert an integer to a float.Even if not converted it wont cause a problem because python by default understands the data type of a value

In [39]:
num = int(input())
ans = 0
for i in range(1,num+1):
    ans += i/(i+1)
print(round(ans,2))

5
3.55


# Question 60
Question:  
Write a program to compute: $f(n)=f(n-1)+100$ when n>0 and $f(0)=0$ with a given n input by console (n>0).

Example: If the following n is given as input to the program: 5

Then, the output of the program should be: 500

In case of input data being supplied to the question, it should be assumed to be a console input.

Hints
We can define recursive function in Python.

In [44]:
def f(n):
    ans = 0
    if n == 0:
        return ans
    return ans + f(n-1) + 100

f(5)

500

In [46]:
def f(n):
    if n == 0:
        return 0
    return f(n-1) + 100

f(5)

500

# Question 61
Question
The Fibonacci Sequence is computed based on the following formula:

$f(n)=0$ if n=0

$f(n)=1$ if n=1

$f(n)=f(n-1)+f(n-2)$ if n>1

Please write a program to compute the value of f(n) with a given n input by console.

Example: If the following n is given as input to the program: 7

Then, the output of the program should be: 13

In case of input data being supplied to the question, it should be assumed to be a console input.

Hints
We can define recursive function in Python.

In [48]:
def f(n):
    if n == 0:
        return 0
    if n == 1:
        return 1
    return f(n-1) + f(n-2)
f(7)

13

In [49]:
def f(n):
    if n < 2:
        return n
    return f(n-1) + f(n-2)
f(7)

13

# Question 62
Question
The Fibonacci Sequence is computed based on the following formula:

$f(n)=0$ if n=0

$f(n)=1$ if n=1

$f(n)=f(n-1)+f(n-2)$ if n>1

Please write a program to compute the value of f(n) with a given n input by console.

Example: If the following n is given as input to the program: 7

Then, the output of the program should be: 0,1,1,2,3,5,8,13

In case of input data being supplied to the question, it should be assumed to be a console input.

Hints
We can define recursive function in Python. Use list comprehension to generate a list from an existing list. Use string.join() to join a list of strings.

In [63]:
def f(n):
    if n == 0:
        return 0
    if n == 1:
        return 1
    return f(n-1) + f(n-2)

lst = [str(f(i)) for i in range(8)]
lst
print(','.join(lst))

0,1,1,2,3,5,8,13


# Question 63
Question
Please write a program using generator to print the even numbers between 0 and n in comma separated form while n is input by console.

Example: If the following n is given as input to the program: 10

Then, the output of the program should be: 0,2,4,6,8,10

In case of input data being supplied to the question, it should be assumed to be a console input.

Hints
Use yield to produce the next value in generator.

In [65]:
def EvenNum(n):
    for i in range(n+1):
        if i % 2 == 0:
            yield i
n = int(input())
print(*EvenNum(n))

10
0 2 4 6 8 10


In [67]:
n = int(input())
for i in range(0,n+1,2):
    if i % 2 == 0:
        print(i,end=',')

10
0,2,4,6,8,10,

# Question 64 - generator/yield
Question:
Please write a program using generator to print the numbers which can be divisible by 5 and 7 between 0 and n in comma separated form while n is input by console.

Example: If the following n is given as input to the program: 100

Then, the output of the program should be: 0,35,70

In case of input data being supplied to the question, it should be assumed to be a console input.

Hints
Use yield to produce the next value in generator.

In [70]:
def generator(n):
    for i in range(n+1):
        if i % 5 == 0 and i % 7 == 0: #i % 35 == 0
            yield i
print(*generator(int(input())))

100
0 35 70


# Question 65 - assert
Question
Please write assert statements to verify that every number in the list [2,4,6,8] is even.

Hints
Use "assert expression" to make assertion.

In [75]:
lst = [2,4,6,8]
for i in lst:
    assert i % 2 == 0, "{} is not an even number".format(i)

In [76]:
data = [2, 4, 5, 6]
for i in data:
    assert i % 2 == 0, "{} is not an even number".format(i)

AssertionError: 5 is not an even number

# Question 66 - eval()
Question:
Please write a program which accepts basic mathematic expression from console and print the evaluation result.

Example: If the following n is given as input to the program: 35 + 3

Then, the output of the program should be: 38

Hints
Use eval() to evaluate an expression.



In [78]:
expression = input()
print(eval(expression))

35 + 3
38


# Question 67 - Binary Search
Question:
Please write a binary search function which searches an item in a sorted list. The function should return the index of element to be searched in the list.

Hints:
Use if/elif to deal with conditions.

In [80]:
def BinarySearch(lst,num):
    l, r = 0, len(lst)
    while l <= r:
        mid = (l + r)//2
        if num == lst[mid]:
            return mid
        elif num < lst[mid]:
            r = mid - 1
        else:
            l = mid + 1
BinarySearch([1,3,4,5,8,12],8)

4

# Question 68 - random.random() -  [0,1] / random.unifrom(start,end)
Question:
Please generate a random float where the value is between 10 and 100 using Python module.

Hints:
Use random.random() to generate a random float in [0,1].

In [85]:
import random
print(random.uniform(10,100))
print(random.randrange(10,100))
print(random.random())

78.61845713113767
46
0.5879403806187777


# Question 69
Question
Please generate a random float where the value is between 5 and 95 using Python module.

Hints
Use random.random() to generate a random float in [0,1].

In [87]:
print(random.uniform(5,95))
print(random.random())

38.80495367649641
0.35427584183512795


# Question 70 - random.choice()
Question:
Please write a program to output a random even number between 0 and 10 inclusive using random module and list comprehension.

Hints:
Use random.choice() to a random element from a list.

In [88]:
print(random.randrange(0,10,2))

2


In [89]:
lst = [i for i in range(0,11,2)]
print(random.choice(lst))

6


# Question 71
Question
Please write a program to output a random number, which is divisible by 5 and 7, between 10 and 150 inclusive using random module and list comprehension.

Hints
Use random.choice() to a random element from a list.



In [93]:
lst = [i for i in range(10,151) if i%35==0]
print(random.choice(lst))

35


# Question 72 - random.sample()
Question
Please write a program to generate a list with 5 random numbers between 100 and 200 inclusive.

Hints
Use random.sample() to generate a list of random values.

In [98]:
print(random.sample(range(100,201),5))

[152, 172, 124, 165, 116]


# Question 73
Question
Please write a program to randomly generate a list with 5 even numbers between 100 and 200 inclusive.

Hints
Use random.sample() to generate a list of random values.

In [99]:
print(random.sample(range(100,201,2),5))

[138, 140, 104, 112, 102]


# Question 74
Question
Please write a program to randomly generate a list with 5 numbers, which are divisible by 5 and 7 , between 1 and 1000 inclusive.

Hints
Use random.sample() to generate a list of random values.

In [101]:
print(random.sample(range(0,1001,35),5))

[770, 350, 595, 420, 245]


In [102]:
lst = [i for i in range(1, 1001) if i % 35 == 0]
resp = random.sample(lst, 5)
print(resp)

[245, 875, 175, 420, 70]


# Question 75
Question
Please write a program to randomly print a integer number between 7 and 15 inclusive.

Hints
Use random.randrange() to a random integer in a given range.

In [107]:
print(random.randint(7,16))
print(random.randrange(7,16))

10
8


# Question 76
Question
Please write a program to compress and decompress the string "hello world!hello world!hello world!hello world!".

Hints
Use zlib.compress() and zlib.decompress() to compress and decompress a string.

In [111]:
import zlib

t = zlib.compress(b"hello world!hello world!hello world!hello world!")
print(t)
print(zlib.decompress(t))

b'x\x9c\xcbH\xcd\xc9\xc9W(\xcf/\xcaIQ\xcc \x82\r\x00\xbd[\x11\xf5'
b'hello world!hello world!hello world!hello world!'


# Question 77
Question
Please write a program to print the running time of execution of "1+1" for 100 times.

Hints
Use timeit() function to measure the running time.

In [120]:

import time
start = time.time()
for i in range(100):
    n = i+i
end = time.time()

print(end-start)

0.0


In [123]:
import timeit
print(timeit.timeit(stmt="[i+i for i in range(1000)]", number=100000))

8.737945300000007


# Question 78
Question
Please write a program to shuffle and print the list [3,6,7,8].

Hints
Use shuffle() function to shuffle a list.

In [130]:
lst = [3,6,7,8]
random.shuffle(lst)
print(lst)

[6, 3, 8, 7]


# Question 79
Question
Please write a program to generate all sentences where subject is in ["I", "You"] and verb is in ["Play", "Love"] and the object is in ["Hockey","Football"].

Hints
Use list[index] notation to get a element from a list.

In [137]:
subject = ["I", "You"]
verb = ["Play", "Love"]
objects = ["Hockey","Football"]

for i in range(5):
    print(subject[random.randrange(0,2)] + " " + verb[random.randrange(0,2)] + " " + objects[random.randrange(0,2)])

I Love Football
I Love Hockey
I Play Football
I Play Football
You Love Hockey


In [138]:
for sub in subject:
    for v in verb:
        for obj in objects:
            print("{} {} {}".format(sub,v,obj))

I Play Hockey
I Play Football
I Love Hockey
I Love Football
You Play Hockey
You Play Football
You Love Hockey
You Love Football


# Question 80 - filter(func,iter,...)
Question
Please write a program to print the list after removing even numbers in [5,6,77,45,22,12,24].

Hints
Use list comprehension to delete a bunch of element from a list.

In [142]:
lst =  [5,6,77,45,22,12,24]
print(list(filter(lambda i:i%2==1,lst)))

[5, 77, 45]


# Question 81 - list comprehension
Question
By using list comprehension, please write a program to print the list after removing numbers which are divisible by 5 and 7 in [12,24,35,70,88,120,155].

Hints
Use list comprehension to delete a bunch of element from a list.

In [145]:
lst = [12,24,35,70,88,120,155]
print(list(filter(lambda x: x%35==0,lst)))

[35, 70]


# Question 82
Question
By using list comprehension, please write a program to print the list after removing the 0th, 2nd, 4th,6th numbers in [12,24,35,70,88,120,155].

Hints
Use list comprehension to delete a bunch of element from a list. Use enumerate() to get (index, value) tuple.

In [146]:
lst = [12,24,35,70,88,120,155]
newLst = [lst[i] for i in range(len(lst)) if i not in [0,2,4,6]]
print(newLst)

[24, 70, 120]


# Question 83
Question
By using list comprehension, please write a program to print the list after removing the 2nd - 4th numbers in [12,24,35,70,88,120,155].

Hints
Use list comprehension to delete a bunch of element from a list. Use enumerate() to get (index, value) tuple.

In [153]:
lst = [12,24,35,70,88,120,155]
print(lst[:2] + lst[5:])
print([lst[i] for i in range(len(lst)) if i < 2 or i > 4])

[12, 24, 120, 155]
[12, 24, 120, 155]


# Question 84
Question
By using list comprehension, please write a program generate a 3x5x8 3D array whose each element is 0.

Hints
Use list comprehension to make an array.

In [157]:
import numpy as np
ma = np.zeros((3,5,8))
print(ma)

[[[0. 0. 0. 0. 0. 0. 0. 0.]
  [0. 0. 0. 0. 0. 0. 0. 0.]
  [0. 0. 0. 0. 0. 0. 0. 0.]
  [0. 0. 0. 0. 0. 0. 0. 0.]
  [0. 0. 0. 0. 0. 0. 0. 0.]]

 [[0. 0. 0. 0. 0. 0. 0. 0.]
  [0. 0. 0. 0. 0. 0. 0. 0.]
  [0. 0. 0. 0. 0. 0. 0. 0.]
  [0. 0. 0. 0. 0. 0. 0. 0.]
  [0. 0. 0. 0. 0. 0. 0. 0.]]

 [[0. 0. 0. 0. 0. 0. 0. 0.]
  [0. 0. 0. 0. 0. 0. 0. 0.]
  [0. 0. 0. 0. 0. 0. 0. 0.]
  [0. 0. 0. 0. 0. 0. 0. 0.]
  [0. 0. 0. 0. 0. 0. 0. 0.]]]


In [160]:
lst = [[[0]*8 for col in range(5)] for row in range(3)]
print(lst)

[[[0, 0, 0, 0, 0, 0, 0, 0], [0, 0, 0, 0, 0, 0, 0, 0], [0, 0, 0, 0, 0, 0, 0, 0], [0, 0, 0, 0, 0, 0, 0, 0], [0, 0, 0, 0, 0, 0, 0, 0]], [[0, 0, 0, 0, 0, 0, 0, 0], [0, 0, 0, 0, 0, 0, 0, 0], [0, 0, 0, 0, 0, 0, 0, 0], [0, 0, 0, 0, 0, 0, 0, 0], [0, 0, 0, 0, 0, 0, 0, 0]], [[0, 0, 0, 0, 0, 0, 0, 0], [0, 0, 0, 0, 0, 0, 0, 0], [0, 0, 0, 0, 0, 0, 0, 0], [0, 0, 0, 0, 0, 0, 0, 0], [0, 0, 0, 0, 0, 0, 0, 0]]]


# Question 85
Question
By using list comprehension, please write a program to print the list after removing the 0th,4th,5th numbers in [12,24,35,70,88,120,155].

Hints
Use list comprehension to delete a bunch of element from a list.Use enumerate() to get (index, value) tuple.

In [161]:
lst = [12,24,35,70,88,120,155]
newL = [lst[i] for i in range(len(lst)) if i not in [0,4,5]]
print(newL)

[24, 35, 70, 155]


# Question 86
Question
By using list comprehension, please write a program to print the list after removing the value 24 in [12,24,35,24,88,120,155].

Hints
Use list's remove method to delete a value.



In [163]:
lst = [12,24,35,24,88,120,155]
lst.remove(24) # this will remove only the first occurrence of 24
print(lst)

[12, 35, 24, 88, 120, 155]


# Question 87 set.intersection / &
Question
With two given lists [1,3,6,78,35,55] and [12,24,35,24,88,120,155], write a program to make a list whose elements are intersection of the above given lists.

Hints
Use set() and "&=" to do set intersection operation.

In [167]:
l1 = [1,3,6,78,35,55] 
l2 = [12,24,35,24,88,120,155]
print(set(l1) & set(l2))

{35}


In [168]:
print(set.intersection(set(l1),set(l2)))

{35}


# Question 88
Question
With a given list [12,24,35,24,88,120,155,88,120,155], write a program to print this list after removing all duplicate values with original order reserved.

Hints
Use set() to store a number of values without duplicate.

In [ ]:
li = [12, 24, 35, 24, 88, 120, 155, 88, 120, 155]
for i in li:
    if li.count(i) > 1:
        li.remove(i)
print(li)


def removeDuplicate(li):
    seen = {}  # dictionary
    for item in li:
        if item not in seen:
            seen[item] = True
            yield item

ans = list(removeDuplicate(li))
print(ans)

In [170]:
lst = [12,24,35,24,88,120,155,88,120,155]
print(set(lst))

{35, 12, 155, 24, 88, 120}


# Question 89
Question
Define a class Person and its two child classes: Male and Female. All classes have a method "getGender" which can print "Male" for Male class and "Female" for Female class.

Hints
Use Subclass(Parentclass) to define a child class.

In [173]:
class Person:
    def __init__(self):
        pass
    def getGender(self):
        return "unknown"
    
class Male(Person):
    def getGender(self):
        return "Male"
class Female(Person):
    def getGender(self):
        return "Female"
    
aMale = Male()
aFemale = Female()
print(aMale.getGender())
print(aFemale.getGender())

Male
Female


# Question 90 - dict
Question: Please write a program which count and print the numbers of each character in a string input by console.

Example: If the following string is given as input to the program: abcdefgabc
Then, the output of the program should be:  
a,2  
c,2  
b,2  
e,1  
d,1  
g,1  
f,1  
Hints: Use dict to store key/value pairs. Use dict.get() method to lookup a key with default value.

In [175]:
string = "abcdefgabc"
for i in set(string):
    print(i,string.count(i))

g 1
c 2
b 2
d 1
f 1
e 1
a 2


In [182]:
d = {}
for i in string:
    if i not in d:
        d[i] = 1
    else:
        d[i] += 1
for key in d:
    print("{},{}".format(key,d.get(key)))

a,2
b,2
c,2
d,1
e,1
f,1
g,1


# Question 91
Question
Please write a program which accepts a string from console and print it in reverse order.

Example: If the following string is given as input to the program:*rise to vote sir

Then, the output of the program should be: ris etov ot esir

Hints
Use list[::-1] to iterate a list in a reverse order.



In [186]:
string = "rise to vote sir"
print(string[::-1])
print(''.join(reversed(string)))

ris etov ot esir
ris etov ot esir


# Question 92
Question
Please write a program which accepts a string from console and print the characters that have even indexes.

Example: If the following string is given as input to the program: H1e2l3l4o5w6o7r8l9d  
Then, the output of the program should be: Helloworld  
Hints :Use list[::2] to iterate a list by step 2.

In [188]:
"H1e2l3l4o5w6o7r8l9d"[::2]

'Helloworld'

# Question 93
Question
Please write a program which prints all permutations of [1,2,3]

Hints
Use itertools.permutations() to get permutations of list.

In [193]:
import itertools
perm = itertools.permutations([1,2,3])
print(list(perm))

[(1, 2, 3), (1, 3, 2), (2, 1, 3), (2, 3, 1), (3, 1, 2), (3, 2, 1)]


# Question 94
Question
Write a program to solve a classic ancient Chinese puzzle: We count 35 heads and 94 legs among the chickens and rabbits in a farm. How many rabbits and how many chickens do we have?

Hints
Use for loop to iterate all possible solutions.



In [195]:
def solveProblem(heads,legs):
    # i: chic
    for i in range(heads+1):
        if (i*2 + (heads-i)*4) == legs:
            return i,heads-i
    return "no solution"

solveProblem(35,94)

(23, 12)

# Question 95
Question
Given the participants' score sheet for your University Sports Day, you are required to find the runner-up score. You are given scores. Store them in a list and find the score of the runner-up.

If the following string is given as input to the program: 5 2 3 6 6 5

Then, the output of the program should be: 5

Hints
Make the scores unique and then find 2nd best number



In [203]:
score = "5 2 3 6 6 5"
score = list(map(int,score.split()))
print(sorted(list(set(score)))[-2])


5


# Question 96
Question
You are given a string S and width W. Your task is to wrap the string into a paragraph of width.

If the following string is given as input to the program: ABCDEFGHIJKLIMNOQRSTUVWXYZ 4

Then, the output of the program should be: ABCD EFGH IJKL IMNO QRST UVWX YZ

Hints
Use wrap function of textwrap module

In [206]:
import textwrap


def wrap(string, max_width):
    string = textwrap.wrap(string, max_width)
    string = "\n".join(string)
    return string


if __name__ == "__main__":
    string, max_width = input(), int(input())
    result = wrap(string, max_width)
    print(result)

ABCDEFGHIJKLIMNOQRSTUVWXYZ
4
ABCD
EFGH
IJKL
IMNO
QRST
UVWX
YZ


In [207]:
import textwrap

string = input()
width = int(input())

print(textwrap.fill(string, width))

ABCDEFGHIJKLIMNOQRSTUVWXYZ
4
ABCD
EFGH
IJKL
IMNO
QRST
UVWX
YZ


# Question 97
Question
You are given an integer, N. Your task is to print an alphabet rangoli of size N. (Rangoli is a form of Indian folk art based on creation of patterns.)

Different sizes of alphabet rangoli are shown below:

$size = 3$

----c----

--c-b-c--

c-b-a-b-c

--c-b-c--

----c----

$size = 5$

--------e--------

------e-d-e------

----e-d-c-d-e----

--e-d-c-b-c-d-e--

e-d-c-b-a-b-c-d-e

--e-d-c-b-c-d-e--

----e-d-c-d-e----

------e-d-e------

--------e--------

Hints
First print the half of the Rangoli in the given way and save each line in a list. Then print the list in reverse order to get the rest.



In [16]:
ord('a')
lst = [chr(i+ord('a')) for i in range(3)]
print(lst)
for i in reversed(range(3)):
    print('-'*(2*i),lst[i],'-'*(2*i))

['a', 'b', 'c']
---- c ----
-- b --
 a 


In [30]:
import string

def print_rangoli(size):
    n = size
    alph = string.ascii_lowercase
    width = 4 * n - 3

    ans = []
    for i in range(n):
        left = "-".join(alph[n - i - 1 : n]) #left: a-b-c
        mid = left[-1:0:-1] + left          #left[-1:0:-1]: c-b-
        final = mid.center(width, "-")
        ans.append(final)

    if len(ans) > 1:
        for i in ans[n - 2 :: -1]:
            ans.append(i)
    ans = "\n".join(ans)
    print(ans)

print_rangoli(3)

 c
c- b-c
c-b- a-b-c
----c----
--c-b-c--
c-b-a-b-c
--c-b-c--
----c----


# Question 98
Question
You are given a date. Your task is to find what the day is on that date.

Input: A single line of input containing the space separated month, day and year, respectively, in MM DD YYYY format. 08 05 2015

Output,Output the correct day in capital letters. WEDNESDAY

Hints
Use weekday function of calender module

In [33]:
import calendar
month, day, year = map(int,input().split())
dayId = calendar.weekday(year,month,day)
print(dayId)
print(calendar.day_name[dayId].upper())

08 05 2015
2
WEDNESDAY


# Question 99 - set, ^, symmetric difference
Question
Given 2 sets of integers, M and N, print their symmetric difference in ascending order. The term symmetric difference indicates those values that exist in either M or N but do not exist in both.

Input

The first line of input contains an integer, M.The second line contains M space-separated integers.The third line contains an integer, N.The fourth line contains N space-separated integers.

4 2 4 5 9 4 2 4 11 12

Output

Output the symmetric difference integers in ascending order, one per line.

5 9 11 12

Hints
Use \'^\' to make symmetric difference operation.

In [40]:
n = input().split()
set1 = set(map(int,n))

m = input().split()
set2 = set(map(int, m))

ans = list(set1 ^ set2)
ans.sort()
for i in ans:
    print(i)

4 2 4 5 9 4 2 4 11 12
5 9 11 12
2
4


# Question 100
Question
You are given words. Some words may repeat. For each word, output its number of occurrences. The output order should correspond with the input order of appearance of the word. See the sample input/output for clarification.

If the following string is given as input to the program:

4 bcdef abcdefg bcde bcdef

Then, the output of the program should be:

3 2 1 1

Hints
Make a list to get the input order and a dictionary to count the word frequency

In [57]:
n = int(input())
word_list = []
word_dict = {}

for i in range(n):
    word = input()
    if word not in word_dict:
        word_list.append(word)
    word_dict[word] = word_dict.get(word, 0) + 1

print(len(word_list))
for word in word_list:
    print(word_dict[word]) 

4
bcdef
abcdefg
bcde
bcdef
3
2
1
1


# Question 101
Question
You are given a string.Your task is to count the frequency of letters of the string and print the letters in descending order of frequency.

If the following string is given as input to the program:

aabbbccde

Then, the output of the program should be:

b 3 a 2 c 2 d 1 e 1

Hints
Count frequency with dictionary and sort by Value from dictionary Items

In [70]:
string = "aabbbccde"
import collections
d = collections.Counter(string)
d = sorted(d.items(),key=lambda x:-x[1])
for item in d:
    print(item[0],item[1],end=' ')

b 3 a 2 c 2 d 1 e 1 

# Question 102
Question
Write a Python program that accepts a string and calculate the number of digits and letters.

Input: Hello321Bye360

Output: Digit - 6,Letter - 8

Hints
Use isdigit() and isalpha() function

In [73]:
Digit, Letter = 0, 0
for i in 'Hello321Bye360':
    if i.isdigit():
        Digit += 1
    elif i.isalpha():
        Letter += 1
print("Digit - {}".format(Digit))
print("Letter - {}".format(Letter))

Digit - 6
Letter - 8


# Question 103 - recursion
Question
Given a number N.Find Sum of 1 to N Using Recursion

Input: 5

Output:15

Hints
Make a recursive function to get the sum

In [76]:
N = int(input())
def sum(N):
    ans = 0
    if N == 0:
        return 0
    return N + sum(N-1)
print(sum(N))

5
15
